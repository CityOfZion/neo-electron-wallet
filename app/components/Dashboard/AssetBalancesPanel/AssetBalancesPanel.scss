@import '../../../styles/animations';

.assetBalancesPanel {
  .centeredContent {
    display: flex;
    justify-content: space-around;
    flex-direction: column;
  }

  .header {
    display: flex;
    align-items: center;
    justify-content: space-between;

    .refresh {
      cursor: pointer;

      &,
      & path {
        color: #9599a2;
        fill: currentColor;
      }

      &.loading {
        animation: spin 2s linear infinite;
      }
    }
  }

  .assets {
    margin: 0 24px;
    display: flex;
    align-items: flex-start;
    justify-content: space-around;
    margin-bottom: 24px;

    .asset {
      text-align: center;

      .quantity {
        height: 48px;
        line-height: 48px;
        color: #000;
        font-size: 48px;
        font-weight: 200;
        margin-bottom: 24px;
      }

      .value {
        color: #969aa0;
        font-size: 14px;
      }

      .change {
        margin-left: 10px;
        color: white;
        font-weight: bold;
        border-radius: 3px;
        padding: 1px 4px;

        &.increase {
          background-color: #5abf6b;
        }

        &.decrease {
          background-color: #ee6d66;
        }
      }
    }
  }

  .totalValue {
    padding: 24px;
    margin: 0 24px 24px;
    background: #f2f2f2;
    text-align: center;
    font-size: 24px;
    font-weight: 500;
  }

  .label {
    margin-bottom: 8px;
    color: #969aa0;
    font-size: 12px;
    font-weight: 300;
    text-transform: uppercase;
  }

  .claim {
    margin: 0 24px;

    .claimButton {
      width: 100%;
    }
  }
}
