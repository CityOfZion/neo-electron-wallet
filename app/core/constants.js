// @flow
export const NETWORK = {
  MAIN: 'MainNet',
  TEST: 'TestNet'
}

export const EXPLORER = {
  NEO_TRACKER: 'Neotracker',
  ANT_CHAIN: 'Antchain'
}

export const ASSETS = {
  NEO: 'NEO',
  GAS: 'GAS'
}

export const ASSETS_LABELS = {
  [ASSETS.NEO]: 'Neo',
  [ASSETS.GAS]: 'Gas'
}

export const BIP44_PATH =
  '8000002C' +
  '80000378' +
  '80000000' +
  '00000000' +
  '00000000'
