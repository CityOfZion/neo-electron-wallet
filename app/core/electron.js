import { shell } from 'electron'

export const openExternal = (srcLink) => {
  shell.openExternal(srcLink)
}
