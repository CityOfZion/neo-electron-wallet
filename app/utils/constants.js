const explorers = [
  {
    id: 'neotracker',
    name: 'NEO Tracker',
    main: 'https://neotracker.io',
    test: 'https://testnet.neotracker.io',
    getTransactionLink(net, txid) {
      const base = net === 'MainNet' ? this.main : this.test;
      return base + '/tx/' + txid;
    },
  },
  {
    id: 'antchain',
    name: 'Antchain',
    main: 'http://antcha.in',
    test: 'http://testnet.antcha.in',
    getTransactionLink(net, txid) {
      const base = net === 'MainNet' ? this.main : this.test;
      return base + '/tx/hash/' + txid;
    },
  },
];
const idToExplorer = {};
explorers.forEach(explorer => {
  idToExplorer[explorer.id] = explorer;
});

export const BLOCKCHAIN_EXPLORERS = idToExplorer;
