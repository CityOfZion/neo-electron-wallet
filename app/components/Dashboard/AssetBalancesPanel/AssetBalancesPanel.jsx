// @flow
import React from 'react'
import classNames from 'classnames'

import Panel from '../../Panel'
import Claim from '../../../containers/Claim'
import Tooltip from '../../../components/Tooltip'
import { formatGAS, formatFiat, formatNEO } from '../../../core/formatters'
import { toNumber, toBigNumber } from '../../../core/math'
import { ASSETS, CURRENCIES } from '../../../core/constants'
import RefreshIcon from '../../../assets/icons/refresh.svg'
import styles from './AssetBalancesPanel.scss'

type Props = {
  className: ?string,
  NEO: string,
  GAS: string,
  neoPrice: number,
  gasPrice: number,
  currencyCode: string,
  refresh: Function
}

export default class AssetBalancesPanel extends React.Component<Props> {
  render = () => {
    const { NEO, GAS, className } = this.props

    return (
      <Panel className={classNames(styles.assetBalancesPanel, className)} renderHeader={this.renderHeader}>
        <div id="balance" className={styles.assets}>
          <div className={styles.asset}>
            <div className={styles.label}>{ASSETS.NEO}</div>
            <div className={styles.quantity} id="amountNeo">{formatNEO(NEO)}</div>
            <div className={styles.value} id="neoWalletValue">
              {this.getFormattedFiatBalance(this.getNEOValue())}
            </div>
          </div>
          <div className={styles.asset}>
            <div className={styles.label}>{ASSETS.GAS}</div>
            <div className={styles.quantity} id="amountGas">
              <Tooltip title={formatGAS(GAS)} disabled={toBigNumber(GAS).eq(0)}>
                {formatGAS(GAS, true)}
              </Tooltip>
            </div>
            <div className={styles.value} id="gasWalletValue">
              {this.getFormattedFiatBalance(this.getGASValue())}
            </div>
          </div>
        </div>
        <div className={styles.totalValue}>
          <div className={styles.label}>Total</div>
          <div id="walletTotal">{this.getFormattedFiatBalance(this.getTotalValue())}</div>
        </div>
        <div className={styles.claim}>
          <Claim className={styles.claimButton} />
        </div>
      </Panel>
    )
  }

  renderHeader = () => {
    return (
      <div className={styles.header}>
        <span>Balances</span>
        <Tooltip title="Refresh">
          <RefreshIcon id="refresh" className={styles.refresh} onClick={this.props.refresh} />
        </Tooltip>
      </div>
    )
  }

  getFormattedFiatBalance = (value: number): string => {
    const { symbol } = CURRENCIES[this.props.currencyCode]
    return `${symbol}${formatFiat(value)}`
  }

  getNEOValue = (): number => {
    const { NEO, neoPrice } = this.props
    return neoPrice && NEO !== '0' ? neoPrice * toNumber(NEO) : 0
  }

  getGASValue = (): number => {
    const { GAS, gasPrice } = this.props
    return gasPrice && GAS !== '0' ? gasPrice * toNumber(GAS) : 0
  }

  getTotalValue = (): number => {
    return this.getNEOValue() + this.getGASValue()
  }
}
