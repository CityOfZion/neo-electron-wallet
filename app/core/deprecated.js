// @flow
import { get } from 'lodash'

import { ID as NETWORK_ID } from '../actions/networkActions'
import { getNetworks } from '../core/networks'

const PREFIX = 'api'

// TODO: Module functions like `doClaimNotify` should be refactored into function props exposed by
//       higher-order components.  Data normally provided by functions like `getNetwork` can instead
//       be retrieved via the `withData` higher-order component.
export const getNetworkId = (state: Object) => {
  return get(state, `${PREFIX}.${NETWORK_ID}.data`)
}

export const getNetwork = (state: Object) => {
  const selectedNetworkId = getNetworkId(state)
  const networks = getNetworks()
  const networkItem = networks.find(({ id, value }) => id === selectedNetworkId) || networks[0]

  return networkItem.network
}

export const getNetworkById = (networkId: string) => {
  const networks = getNetworks()
  const networkItem = networks.find(({ id, value }) => id === networkId) || networks[0]
  return networkItem.network
}
