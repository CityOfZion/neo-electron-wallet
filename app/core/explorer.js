// @flow
import { openExternal } from './electron'
import explorerConfig from './explorerConfig'
import { isMainNetwork } from './networks'

export const buildExplorerLink = (
  networkId: string,
  explorer: ExplorerType,
  structure: string,
  suffix: string
) => {
  if (Object.prototype.hasOwnProperty.call(explorerConfig, explorer)) {
    const explorerInfo = explorerConfig[explorer]

    return getExplorerBaseURL(networkId, explorer) + explorerInfo[structure] + suffix
  }

  return ''
}

export const getExplorerBaseURL = (networkId: string, explorer: ExplorerType) => {
  if (Object.prototype.hasOwnProperty.call(explorerConfig, explorer)) {
    const selectedExplorer = explorerConfig[explorer]

    return isMainNetwork(networkId)
      ? selectedExplorer.mainNetwork
      : selectedExplorer.testNetwork
  }
}

export const getExplorerTxLink = (networkId: string, explorer: ExplorerType, txId: string) => buildExplorerLink(networkId, explorer, 'trxLinkStructure', txId)

export const getExplorerAddressLink = (networkId: string, explorer: ExplorerType, address: string) => buildExplorerLink(networkId, explorer, 'addressStructure', address)

export const getExplorerAssetLink = (networkId: string, explorer: ExplorerType, assetId: string) => buildExplorerLink(networkId, explorer, 'assetLinkStructure', assetId)

export const openExplorerTx = (networkId: string, explorer: ExplorerType, txId: string) => openExternal(getExplorerTxLink(networkId, explorer, txId))

export const openExplorerAddress = (networkId: string, explorer: ExplorerType, address: string) => openExternal(getExplorerAddressLink(networkId, explorer, address))

export const openExplorerAsset = (networkId: string, explorer: ExplorerType, assetId: string) => openExternal(getExplorerAssetLink(networkId, explorer, assetId))
