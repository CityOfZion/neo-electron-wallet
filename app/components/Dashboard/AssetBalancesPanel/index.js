// @flow
import { compose } from 'recompose'
import { values, omit } from 'lodash'
import { createBatchActions, withActions } from 'spunky'

import AssetBalancesPanel from './AssetBalancesPanel'
import balancesActions from '../../../actions/balancesActions'
import claimsActions from '../../../actions/claimsActions'
import pricesActions from '../../../actions/pricesActions'
import withBalancesData from '../../../hocs/withBalancesData'
import withCurrencyData from '../../../hocs/withCurrencyData'
import withPricesData from '../../../hocs/withPricesData'
import withLoadingProp from '../../../hocs/withLoadingProp'
import withProgressPanel from '../../../hocs/withProgressPanel'
import withSuccessNotification from '../../../hocs/withSuccessNotification'
import withFailureNotification from '../../../hocs/withFailureNotification'

const mapBalanceDataToProps = (balances) => ({
  NEO: balances.NEO,
  GAS: balances.GAS,
  tokenBalances: values(omit(balances, 'NEO', 'GAS'))
})

const mapPricesDataToProps = ({ NEO, GAS }) => ({
  neoPrice: NEO,
  gasPrice: GAS
})

const mapBalancesActionsToProps = (actions, props) => ({
  refresh: () => actions.call({ net: props.net, address: props.address, tokens: props.tokens })
})

// TODO: move this into its own actions file
const batchActions = createBatchActions('assetBalancesPanel', {
  prices: pricesActions,
  claims: claimsActions,
  balances: balancesActions
})

export default compose(
  withCurrencyData('currencyCode'),

  // Fetch price & balance data based based upon the selected currency.
  // Reload data with the currency changes.
  withProgressPanel(batchActions, { title: 'Balances' }),
  withPricesData(mapPricesDataToProps),
  withBalancesData(mapBalanceDataToProps),

  // Expose data & functionality needed for `refresh` action.
  withActions(balancesActions, mapBalancesActionsToProps),
  withLoadingProp(balancesActions),
  withSuccessNotification(balancesActions, 'Received latest blockchain information.'),
  withFailureNotification(balancesActions, 'Failed to retrieve blockchain information.')
)(AssetBalancesPanel)
