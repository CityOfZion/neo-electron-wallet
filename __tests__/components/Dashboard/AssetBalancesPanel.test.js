import React from 'react'
import { merge } from 'lodash'
import { mount, shallow } from 'enzyme'

import { createStore, provideStore, provideState } from '../../testHelpers'
import AssetBalancesPanel from '../../../app/components/Dashboard/AssetBalancesPanel'
import { DEFAULT_CURRENCY_CODE, MAIN_NETWORK_ID } from '../../../app/core/constants'
import { LOADED } from '../../../app/values/state'

// TODO research how to move the axios mock code which is repeated in NetworkSwitch to a helper or config file
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'

const axiosMock = new MockAdapter(axios)
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/?limit=0&convert=USD')
  .reply(200, [{ symbol: 'NEO', price_usd: 24.5 }, { symbol: 'GAS', price_usd: 18.2 }])

const initialState = {
  api: {
    NETWORK: {
      batch: false,
      state: LOADED,
      data: MAIN_NETWORK_ID
    },
    AUTH: {
      batch: false,
      state: LOADED,
      data: {
        address: 'ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw'
      }
    },
    SETTINGS: {
      batch: false,
      state: LOADED,
      data: {
        currency: DEFAULT_CURRENCY_CODE,
        tokens: []
      }
    },
    PRICES: {
      batch: false,
      state: LOADED,
      data: {
        NEO: 25.48,
        GAS: 18.1
      }
    },
    BALANCES: {
      batch: false,
      state: LOADED,
      data: {
        NEO: '100001',
        GAS: '1000.0001601'
      }
    },
    CLAIMS: {
      batch: false,
      state: LOADED,
      data: {
        total: '0.5'
      }
    }
  },
  claim: {
    claimRequest: false,
    disableClaimButton: false
  }
}

describe('AssetBalancesPanel', () => {
  test('renders without crashing', () => {
    const store = createStore(initialState)
    const wrapper = shallow(<AssetBalancesPanel store={store} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('correctly renders data from state', () => {
    const wrapper = mount(provideState(<AssetBalancesPanel />, initialState))

    expect(wrapper.find('#neoWalletValue').text()).toEqual(`$2,548,025.48`)
    expect(wrapper.find('#gasWalletValue').text()).toEqual(`$18,100.00`)
    expect(wrapper.find('#walletTotal').text()).toEqual(`$2,566,125.48`)
    expect(wrapper.find('#amountNeo').text()).toEqual('100,001')
    expect(wrapper.find('#amountGas').text()).toEqual('1,000.0002')
  })

  test('account data refreshes when refresh button is clicked', () => {
    const store = createStore(initialState)
    const wrapper = mount(provideStore(<AssetBalancesPanel />, store))

    wrapper.find('#refresh').simulate('click')

    expect(store.getActions()).toContainEqual(expect.objectContaining({
      type: 'BALANCES/REQ/REQUEST',
      meta: expect.objectContaining({ id: 'BALANCES' })
    }))
  })

  test('correctly renders data from state with non-default currency', () => {
    const state = merge(initialState, {
      api: {
        SETTINGS: { data: { currency: 'eur' } },
        PRICES: { data: { NEO: 1.11, GAS: 0.55 } }
      }
    })
    const wrapper = mount(provideState(<AssetBalancesPanel />, state))

    expect(wrapper.find('#neoWalletValue').text()).toEqual(`€111,001.11`)
    expect(wrapper.find('#gasWalletValue').text()).toEqual(`€550.00`)
    expect(wrapper.find('#walletTotal').text()).toEqual(`€111,551.11`)
  })
})
