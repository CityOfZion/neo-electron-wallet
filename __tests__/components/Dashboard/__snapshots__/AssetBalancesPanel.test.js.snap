// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetBalancesPanel renders without crashing 1`] = `
<withData(Connect(withData(Connect(withData(Connect(withData(withProps(Connect(withData(Connect(withData(withProps(Connect(withData(Connect(withActions(Connect(Connect(withProgressProp(LoadedNotifier))))))))))))))))))))
  dispatch={[Function]}
  gasPrice={18.1}
  neoPrice={25.48}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
